File: sliding_window/Longest Substring Without Repeating Characters.go
package main

import (
	"fmt"
)

func main(){
	// s := "abcabcbb"
	s := "pwwkew"
	fmt.Println(s)

	bucket := map[byte]bool{}
	start := 0
	maxLength := 0

	for end:=0; end<len(s); end++{
		for bucket[s[end]]{
			bucket[s[start]] = false
			start++
		}
		bucket[s[end]] = true

		if end - start + 1 > maxLength{
			maxLength = end-start+1
		}
	}

	fmt.Println(maxLength)
}

func findMax(a,b int)int{
	if a>b{
		return a
	}
	return b
}

File: sliding_window/Longest Substring Without Repeating Characters.py
s = "abcabcbb"

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        bucket = set()
        max_len = 0
        start, end = 0, 0

        while end < len(s):
            if s[end] not in bucket:
                bucket.add(s[end])
                end += 1
            
            else:
                while s[end] in bucket:
                    bucket.remove(s[start])
                    start += 1
            
            max_len = max(max_len, len(bucket))
        return max_len

sln = Solution()
print(sln.lengthOfLongestSubstring(s))